lib/utils/utils.go

```diff
@@ -512,29 +513,75 @@ func WriteHostUUID(dataDir string, id string) error {
 // ReadOrMakeHostUUID looks for a hostid file in the data dir. If present,
 // returns the UUID from it, otherwise generates one
 func ReadOrMakeHostUUID(dataDir string) (string, error) {
-	id, err := ReadHostUUID(dataDir)
-	if err == nil {
+	hostUUIDFileName := GetHostUUIDPath(dataDir)
+	hostUUIDLock := hostUUIDFileName + ".lock"
+
+	iterationLimit := 3
+
+	for i := 0; i < iterationLimit; i++ {
+		id, err := ReadHostUUID(dataDir)
+		if err == nil {
+			return id, nil
+		}
+		if !trace.IsNotFound(err) {
+			return "", trace.Wrap(err)
+		}
+
+		// Checking error instead of the usual uuid.New() in case uuid generation
+		// fails due to not enough randomness. It's been known to happen happen when
+		// Teleport starts very early in the node initialization cycle and /dev/urandom
+		// isn't ready yet.
+		rawID, err := uuid.NewRandom()
+		if err != nil {
+			return "", trace.BadParameter("" +
+				"Teleport failed to generate host UUID. " +
+				"This may happen if randomness source is not fully initialized when the node is starting up. " +
+				"Please try restarting Teleport again.")
+		}
+
+		id = rawID.String()
+
+		writeFile := func() (string, error) {
+			f, err := os.OpenFile(hostUUIDLock, os.O_WRONLY|os.O_CREATE, 0o400)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+
+			unlock, err := FSTryWriteLock(f.Name())
```

Comment:
I believe that `FSTryWriteLock` will create (and clean up?) the lock file on its own, so there's no need to create it or delete it here - moreover, we should not delete the lockfile if there's been errors, I think?